### PR TITLE
Fix pad_to_max_length Whisper

### LINKED ISSUE #30223: Trying to stack tensors from different devices in `_pad_to_max_length` in Whisper batched inference
This issue seems to be due to the following line, added in #29065 to fix #29036, but the fix doesn't work with batched inference on GPU/MPS because the tensor is on the wrong device:
https://github.com/huggingface/transformers/blob/4f7b434acbebd8ef552615ea6c81df4bd9b57813/src/transformers/models/whisper/generation_whisper.py#L146

### System Info

- `transformers` version: 4.40.0.dev0 bf9a7ab
- Platform: macOS-14.2.1-arm64-arm-64bit
- Python version: 3.10.13
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: 0.28.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.2.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@ylacombe @sanchit-gandhi

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from datasets import Audio, load_dataset
from transformers import WhisperForConditionalGeneration, AutoProcessor
import torch
import numpy as np

model = WhisperForConditionalGeneration.from_pretrained(
    "openai/whisper-tiny", torch_dtype=torch.float16
)
processor = AutoProcessor.from_pretrained("openai/whisper-tiny")
model.to("mps")

ds = load_dataset("distil-whisper/meanwhile", "default")["test"]
ds = ds.cast_column("audio", Audio(sampling_rate=16000))
audio = ds[:8]["audio"]
audio = [x["array"] for x in audio]
audio[0][:] = np.random.normal(scale=0.05, size=audio[0].shape)
inputs = processor(
    audio,
    return_tensors="pt",
    truncation=False,
    padding="longest",
    return_attention_mask=True,
    sampling_rate=16_000,
)
inputs = inputs.to(model.device, torch.float16)

result = model.generate(
    **inputs,
    no_speech_threshold=0.2,
    logprob_threshold=0.0,
    temperature=(0.0,),
    task="transcribe",
    language="fr",
)
decoded = processor.batch_decode(
    result, skip_special_tokens=False, decode_with_timestamps=True
)
print(decoded)
```
```pytb
Traceback (most recent call last):
  File "/Users/ondra/bordel/test-hf/repr_batch_device_issue.py", line 27, in <module>
    result = model.generate(
  File "/Users/ondra/mambaforge/envs/transformers/lib/python3.10/site-packages/transformers/models/whisper/generation_whisper.py", line 730, in generate
    sequences = _pad_to_max_length(final_segments, generation_config.pad_token_id, padding="right")
  File "/Users/ondra/mambaforge/envs/transformers/lib/python3.10/site-packages/transformers/models/whisper/generation_whisper.py", line 153, in _pad_to_max_length
    sequences = torch.stack(sequences, dim=0)
RuntimeError: torch.cat(): all input tensors must be on the same device. Received cpu and mps:0
```

### Expected behavior

No error

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?
Fix  #30223!

I've added the `device` argument because one of the method using pad_to_max_length is static !

### CODE REVIEW

File: src/transformers/models/whisper/generation_whisper.py [hunk 1]
```diff
@@ -727,7 +729,9 @@ def generate(
             if (prompt_ids is not None and generation_config.prompt_condition_type == "first-segment")
             else current_segments
         )
-        sequences = _pad_to_max_length(final_segments, generation_config.pad_token_id, padding="right")
+        sequences = _pad_to_max_length(
+            final_segments, generation_config.pad_token_id, device=self.device, padding="right"
```
Comment (amyeroberts) [inline src/transformers/models/whisper/generation_whisper.py:733]: Is self.device guaranteed to be the same device as `final_segments`? What happens if you load the model using `device_map="auto"` in a multi-gpu setting?

File: src/transformers/models/whisper/generation_whisper.py [hunk 2]
```diff
@@ -122,7 +122,9 @@ def _get_attr_from_logit_processors(logits_processor, logit_processor_class, att
     return None
 
 
-def _pad_to_max_length(current_segments, pad_token_id, padding="right", bos_token_tensor=None, cut_off_length=None):
+def _pad_to_max_length(
+    current_segments, pad_token_id, device, padding="right", bos_token_tensor=None, cut_off_length=None
```
Comment (amyeroberts) [inline src/transformers/models/whisper/generation_whisper.py:126]: Do we need to add the device argument? Couldn't we take it from `current_segments`?

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this! 

Some Qs about the device assumption. Could you add a test?

File: src/transformers/models/whisper/generation_whisper.py (hunk 2 above)
Comment (ylacombe) [reply src/transformers/models/whisper/generation_whisper.py:126]: I thought about it, but there might be some edge cases in which `current_segments` is empty, so I'd rather add a device argument. 
Additionally, `device`  was already passed to `_prepare_decoder_input_ids`, which is a static method that calls `_pad_to_max_length`, so it isn't far-stretched to also pass it to the method IMO!

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
Comment (ylacombe) [reply src/transformers/models/whisper/generation_whisper.py:739]: Not sure how `auto` device works, but on my set-up it works well, adding some tests rn!

**Comment by ylacombe** [comment]
Hey @amyeroberts, I'm not totally sure about how to properly test auto device mapping, but I've added two tests, one on a single gpu settings, the other on the multi-gpu settings, does it work for you ?
Both were tested and passed on a DGX !

File: src/transformers/models/whisper/generation_whisper.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/models/whisper/generation_whisper.py:126]: Sounds good to me - thanks for providing the extra context re `_prepare_decoder_input_ids`

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing and for the extensive tests!

**Comment by ylacombe** [comment]
Thanks @amyeroberts ! I still need your approval to merge!

**Review by amyeroberts — APPROVED** [review]
Woops - sorry, I meant to approve before!

**Pull request merged**